File: src/main/services/store.ts
import { existsSync } from 'node:fs'
import { mkdir, readFile, rename, writeFile } from 'node:fs/promises'
import { createRequire } from 'node:module'
import { dirname, join } from 'node:path'
import initSqlJs, { type Database } from 'sql.js'
import type { DanmakuFilter, DanmakuRecord, LogEntry, Rule } from '@shared/types'

type SqlRow = Record<string, unknown>
const localRequire = createRequire(import.meta.url)

export class SqliteStore {
  private db!: Database

  async init(filePath: string): Promise<void> {
    await mkdir(dirname(filePath), { recursive: true })
    const wasmPath = this.resolveWasmPath()
    const SQL = await initSqlJs({ locateFile: () => wasmPath })
    if (existsSync(filePath)) {
      const bytes = await readFile(filePath)
      this.db = new SQL.Database(new Uint8Array(bytes))
    } else {
      this.db = new SQL.Database()
    }
    this.db.run('PRAGMA journal_mode = WAL;')
    this.db.run(`
      CREATE TABLE IF NOT EXISTS rules (
        id TEXT PRIMARY KEY,
        json TEXT NOT NULL,
        updated_at INTEGER NOT NULL
      );
      CREATE TABLE IF NOT EXISTS danmaku_records (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        source TEXT NOT NULL,
        room_id TEXT,
        kind TEXT NOT NULL,
        user_id TEXT,
        user_name TEXT,
        text TEXT,
        gift_name TEXT,
        gift_count INTEGER,
        gift_value REAL,
        repeat_count INTEGER,
        ts INTEGER NOT NULL,
        created_at INTEGER NOT NULL,
        matched_rule_id TEXT,
        matched_rule_name TEXT,
        action_result TEXT NOT NULL DEFAULT 'none',
        fail_reason TEXT,
        raw_json TEXT
      );
      CREATE INDEX IF NOT EXISTS idx_danmaku_ts ON danmaku_records(ts);
      CREATE INDEX IF NOT EXISTS idx_danmaku_source_ts ON danmaku_records(source, ts);
      CREATE INDEX IF NOT EXISTS idx_danmaku_kind_ts ON danmaku_records(kind, ts);
      CREATE INDEX IF NOT EXISTS idx_danmaku_user ON danmaku_records(user_name);
      CREATE TABLE IF NOT EXISTS logs (
        id INTEGER PRIMARY KEY,
        level TEXT NOT NULL,
        category TEXT NOT NULL,
        message TEXT NOT NULL,
        detail TEXT,
        ts INTEGER NOT NULL
      );
      CREATE TABLE IF NOT EXISTS settings (
        key TEXT PRIMARY KEY,
        value TEXT NOT NULL
      );
    `)
    await this.persist(filePath)
  }

  async listRules(): Promise<Rule[]> {
    return this.all('SELECT json FROM rules ORDER BY json_extract(json, \'$.priority\') DESC, updated_at ASC')
      .map((row) => JSON.parse(String(row.json)) as Rule)
  }

  async saveRule(rule: Rule): Promise<Rule> {
    const updated = { ...rule, updatedAt: Date.now(), createdAt: rule.createdAt ?? Date.now() }
    this.run('INSERT OR REPLACE INTO rules(id, json, updated_at) VALUES (?, ?, ?)', [updated.id, JSON.stringify(updated), updated.updatedAt])
    return updated
  }

  async removeRule(id: string): Promise<void> {
    this.run('DELETE FROM rules WHERE id = ?', [id])
  }

  async clearRules(): Promise<void> {
    this.db.run('DELETE FROM rules')
  }

  insertRecord(record: Omit<DanmakuRecord, 'id'>): number {
    this.run(
      `INSERT INTO danmaku_records
        (source, room_id, kind, user_id, user_name, text, gift_name, gift_count, gift_value, repeat_count, ts, created_at, matched_rule_id, matched_rule_name, action_result, fail_reason, raw_json)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        record.source,
        record.roomId ?? null,
        record.kind,
        record.userId ?? null,
        record.userName ?? null,
        record.text ?? null,
        record.giftName ?? null,
        record.giftCount ?? null,
        record.giftValue ?? null,
        record.repeatCount ?? null,
        record.ts,
        record.createdAt,
        record.matchedRuleId ?? null,
        record.matchedRuleName ?? null,
        record.actionResult,
        record.failReason ?? null,
        record.rawJson ?? null,
      ],
    )
    const row = this.all('SELECT last_insert_rowid() AS id')[0]
    return Number(row?.id ?? 0)
  }

  updateRecordResult(id: number, result: Pick<DanmakuRecord, 'matchedRuleId' | 'matchedRuleName' | 'actionResult' | 'failReason'>): void {
    this.run('UPDATE danmaku_records SET matched_rule_id = ?, matched_rule_name = ?, action_result = ?, fail_reason = ? WHERE id = ?', [
      result.matchedRuleId ?? null,
      result.matchedRuleName ?? null,
      result.actionResult,
      result.failReason ?? null,
      id,
    ])
  }

  queryRecords(filter: DanmakuFilter = {}): DanmakuRecord[] {
    const { where, params } = buildRecordWhere(filter)
    const limit = Math.min(Math.max(filter.limit ?? 100, 1), 500)
    const offset = Math.max(filter.offset ?? 0, 0)
    return this.all(
      `SELECT id, source, room_id AS roomId, kind, user_id AS userId, user_name AS userName,
        text, gift_name AS giftName, gift_count AS giftCount, gift_value AS giftValue,
        repeat_count AS repeatCount, ts, created_at AS createdAt, matched_rule_id AS matchedRuleId,
        matched_rule_name AS matchedRuleName, action_result AS actionResult, fail_reason AS failReason,
        raw_json AS rawJson FROM danmaku_records ${where} ORDER BY ts DESC, id DESC LIMIT ? OFFSET ?`,
      [...params, limit, offset],
    ).map((row) => normalizeRecord(row))
  }

  countRecords(filter: DanmakuFilter = {}): number {
    const { where, params } = buildRecordWhere(filter)
    const row = this.all(`SELECT COUNT(*) AS count FROM danmaku_records ${where}`, params)[0]
    return Number(row?.count ?? 0)
  }

  clearRecords(range: { from?: number; to?: number } = {}): number {
    const conditions: string[] = []
    const params: unknown[] = []
    if (range.from !== undefined) { conditions.push('ts >= ?'); params.push(range.from) }
    if (range.to !== undefined) { conditions.push('ts <= ?'); params.push(range.to) }
    const before = this.countRecords()
    this.run(`DELETE FROM danmaku_records${conditions.length ? ` WHERE ${conditions.join(' AND ')}` : ''}`, params)
    return before - this.countRecords()
  }

  cleanupRetention(days: number, maxRows: number): number {
    const cutoff = Date.now() - Math.max(days, 1) * 24 * 60 * 60 * 1000
    const before = this.countRecords()
    this.run('DELETE FROM danmaku_records WHERE ts < ?', [cutoff])
    const excess = Math.max(0, this.countRecords() - Math.max(maxRows, 1))
    if (excess) this.run('DELETE FROM danmaku_records WHERE id IN (SELECT id FROM danmaku_records ORDER BY ts ASC, id ASC LIMIT ?)', [excess])
    return before - this.countRecords()
  }

  saveLog(entry: LogEntry): void {
    this.run('INSERT OR REPLACE INTO logs(id, level, category, message, detail, ts) VALUES (?, ?, ?, ?, ?, ?)', [entry.id, entry.level, entry.category, entry.message, entry.detail ?? null, entry.ts])
  }

  listLogs(limit = 100): LogEntry[] {
    return this.all('SELECT id, level, category, message, detail, ts FROM logs ORDER BY id DESC LIMIT ?', [Math.min(Math.max(limit, 1), 500)]) as unknown as LogEntry[]
  }

  getSetting<T>(key: string, fallback: T): T {
    const row = this.all('SELECT value FROM settings WHERE key = ?', [key])[0]
    if (!row) return fallback
    try { return JSON.parse(String(row.value)) as T } catch { return fallback }
  }

  setSetting<T>(key: string, value: T): void {
    this.run('INSERT OR REPLACE INTO settings(key, value) VALUES (?, ?)', [key, JSON.stringify(value)])
  }

  async persist(filePath: string): Promise<void> {
    if (!this.db) return
    const bytes = this.db.export()
    const tempPath = `${filePath}.tmp`
    await writeFile(tempPath, bytes)
    await rename(tempPath, filePath)
  }

  private all(sql: string, params: unknown[] = []): SqlRow[] {
    const statement = this.db.prepare(sql)
    statement.bind(params as any)
    const rows: SqlRow[] = []
    while (statement.step()) rows.push(statement.getAsObject() as SqlRow)
    statement.free()
    return rows
  }

  private run(sql: string, params: unknown[] = []): void {
    const statement = this.db.prepare(sql)
    statement.run(params as any)
    statement.free()
  }

  private resolveWasmPath(): string {
    try {
      const resolved = localRequire.resolve('sql.js/dist/sql-wasm.wasm')
      return resolved
    } catch {
      return join(process.resourcesPath, 'sql.js', 'sql-wasm.wasm')
    }
  }
}

function buildRecordWhere(filter: DanmakuFilter): { where: string; params: unknown[] } {
  const clauses: string[] = []
  const params: unknown[] = []
  if (filter.from !== undefined) { clauses.push('ts >= ?'); params.push(filter.from) }
  if (filter.to !== undefined) { clauses.push('ts <= ?'); params.push(filter.to) }
  if (filter.source) { clauses.push('source = ?'); params.push(filter.source) }
  if (filter.kind) { clauses.push('kind = ?'); params.push(filter.kind) }
  if (filter.userName) { clauses.push('user_name LIKE ?'); params.push(`%${filter.userName}%`) }
  if (filter.keyword) { clauses.push('(text LIKE ? OR gift_name LIKE ?)'); params.push(`%${filter.keyword}%`, `%${filter.keyword}%`) }
  if (filter.actionResult) { clauses.push('action_result = ?'); params.push(filter.actionResult) }
  if (filter.matched === 'yes') clauses.push('matched_rule_id IS NOT NULL')
  if (filter.matched === 'no') clauses.push('matched_rule_id IS NULL')
  return { where: clauses.length ? `WHERE ${clauses.join(' AND ')}` : '', params }
}

function normalizeRecord(row: SqlRow): DanmakuRecord {
  return {
    id: Number(row.id), source: String(row.source), roomId: nullableString(row.roomId), kind: row.kind as DanmakuRecord['kind'],
    userId: nullableString(row.userId), userName: nullableString(row.userName), text: nullableString(row.text),
    giftName: nullableString(row.giftName), giftCount: nullableNumber(row.giftCount), giftValue: nullableNumber(row.giftValue),
    repeatCount: nullableNumber(row.repeatCount), ts: Number(row.ts), createdAt: Number(row.createdAt),
    matchedRuleId: nullableString(row.matchedRuleId), matchedRuleName: nullableString(row.matchedRuleName),
    actionResult: (row.actionResult as DanmakuRecord['actionResult']) ?? 'none', failReason: nullableString(row.failReason), rawJson: nullableString(row.rawJson),
  }
}

function nullableString(value: unknown): string | undefined { return value === null || value === undefined ? undefined : String(value) }
function nullableNumber(value: unknown): number | undefined { return value === null || value === undefined ? undefined : Number(value) }

File: store.go
package main

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	_ "modernc.org/sqlite"
)

type Store struct {
	db *sql.DB
}

func OpenStore(file string) (*Store, error) {
	if err := os.MkdirAll(filepath.Dir(file), 0o700); err != nil {
		return nil, fmt.Errorf("create data directory: %w", err)
	}
	db, err := sql.Open("sqlite", file)
	if err != nil {
		return nil, fmt.Errorf("open sqlite database: %w", err)
	}
	db.SetMaxOpenConns(1)
	db.SetConnMaxLifetime(0)
	store := &Store{db: db}
	if _, err := db.Exec(`PRAGMA busy_timeout=5000; PRAGMA journal_mode=WAL; PRAGMA foreign_keys=ON;`); err != nil {
		db.Close()
		return nil, fmt.Errorf("configure sqlite database: %w", err)
	}
	if err := store.initSchema(context.Background()); err != nil {
		db.Close()
		return nil, err
	}
	return store, nil
}

func (s *Store) Close() error { return s.db.Close() }

func (s *Store) initSchema(ctx context.Context) error {
	_, err := s.db.ExecContext(ctx, `
		CREATE TABLE IF NOT EXISTS rules (
			id TEXT PRIMARY KEY,
			json TEXT NOT NULL,
			updated_at INTEGER NOT NULL
		);
		CREATE TABLE IF NOT EXISTS danmaku_records (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			source TEXT NOT NULL,
			room_id TEXT,
			kind TEXT NOT NULL,
			user_id TEXT,
			user_name TEXT,
			text TEXT,
			gift_name TEXT,
			gift_count INTEGER,
			gift_value REAL,
			repeat_count INTEGER,
			ts INTEGER NOT NULL,
			created_at INTEGER NOT NULL,
			matched_rule_id TEXT,
			matched_rule_name TEXT,
			action_result TEXT NOT NULL DEFAULT 'none',
			fail_reason TEXT,
			raw_json TEXT
		);
		CREATE INDEX IF NOT EXISTS idx_danmaku_ts ON danmaku_records(ts);
		CREATE INDEX IF NOT EXISTS idx_danmaku_source_ts ON danmaku_records(source, ts);
		CREATE INDEX IF NOT EXISTS idx_danmaku_kind_ts ON danmaku_records(kind, ts);
		CREATE INDEX IF NOT EXISTS idx_danmaku_user ON danmaku_records(user_name);
		CREATE TABLE IF NOT EXISTS logs (
			id INTEGER PRIMARY KEY,
			level TEXT NOT NULL,
			category TEXT NOT NULL,
			message TEXT NOT NULL,
			detail TEXT,
			ts INTEGER NOT NULL
		);
		CREATE TABLE IF NOT EXISTS settings (
			key TEXT PRIMARY KEY,
			value TEXT NOT NULL
		);
	`)
	if err != nil {
		return fmt.Errorf("initialize sqlite schema: %w", err)
	}
	return nil
}

func (s *Store) ListRules(ctx context.Context) ([]Rule, error) {
	rows, err := s.db.QueryContext(ctx, `SELECT json FROM rules ORDER BY json_extract(json, '$.priority') DESC, updated_at ASC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	result := make([]Rule, 0)
	for rows.Next() {
		var raw string
		if err := rows.Scan(&raw); err != nil {
			return nil, err
		}
		var rule Rule
		if err := json.Unmarshal([]byte(raw), &rule); err != nil {
			continue
		}
		result = append(result, rule)
	}
	return result, rows.Err()
}

func (s *Store) SaveRule(ctx context.Context, rule Rule) (Rule, error) {
	now := nowMillis()
	if rule.CreatedAt == 0 {
		rule.CreatedAt = now
	}
	rule.UpdatedAt = now
	data, err := json.Marshal(rule)
	if err != nil {
		return Rule{}, err
	}
	_, err = s.db.ExecContext(ctx, `INSERT INTO rules(id,json,updated_at) VALUES(?,?,?) ON CONFLICT(id) DO UPDATE SET json=excluded.json, updated_at=excluded.updated_at`, rule.ID, string(data), now)
	return rule, err
}

func (s *Store) RemoveRule(ctx context.Context, id string) error {
	_, err := s.db.ExecContext(ctx, `DELETE FROM rules WHERE id=?`, id)
	return err
}

func (s *Store) ClearRules(ctx context.Context) error {
	_, err := s.db.ExecContext(ctx, `DELETE FROM rules`)
	return err
}

func (s *Store) InsertRecord(ctx context.Context, record DanmakuRecord) (int64, error) {
	result, err := s.db.ExecContext(ctx, `INSERT INTO danmaku_records
		(source,room_id,kind,user_id,user_name,text,gift_name,gift_count,gift_value,repeat_count,ts,created_at,matched_rule_id,matched_rule_name,action_result,fail_reason,raw_json)
		VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)`,
		record.Source, nullable(record.RoomID), string(record.Kind), nullable(record.UserID), nullable(record.UserName), nullable(record.Text),
		nullable(record.GiftName), nullableInt(record.GiftCount), nullableFloat(record.GiftValue), nullableInt(record.RepeatCount), record.TS, record.CreatedAt,
		nullable(record.MatchedRuleID), nullable(record.MatchedRuleName), record.ActionResult, nullable(record.FailReason), nullable(record.RawJSON))
	if err != nil {
		return 0, err
	}
	return result.LastInsertId()
}

func (s *Store) UpdateRecordResult(ctx context.Context, id int64, outcome RuleOutcome) error {
	_, err := s.db.ExecContext(ctx, `UPDATE danmaku_records SET matched_rule_id=?,matched_rule_name=?,action_result=?,fail_reason=? WHERE id=?`,
		nullable(outcome.RuleID), nullable(outcome.RuleName), outcome.Result, nullable(outcome.Reason), id)
	return err
}

func (s *Store) QueryRecords(ctx context.Context, filter DanmakuFilter) ([]DanmakuRecord, error) {
	where, args := recordWhere(filter)
	limit := filter.Limit
	if limit <= 0 || limit > 500 {
		limit = 100
	}
	offset := max(filter.Offset, 0)
	args = append(args, limit, offset)
	rows, err := s.db.QueryContext(ctx, `SELECT id,source,room_id,kind,user_id,user_name,text,gift_name,gift_count,gift_value,repeat_count,ts,created_at,matched_rule_id,matched_rule_name,action_result,fail_reason,raw_json FROM danmaku_records `+where+` ORDER BY ts DESC,id DESC LIMIT ? OFFSET ?`, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	result := make([]DanmakuRecord, 0)
	for rows.Next() {
		record, err := scanRecord(rows)
		if err != nil {
			return nil, err
		}
		result = append(result, record)
	}
	return result, rows.Err()
}

func (s *Store) CountRecords(ctx context.Context, filter DanmakuFilter) (int, error) {
	where, args := recordWhere(filter)
	var count int
	err := s.db.QueryRowContext(ctx, `SELECT COUNT(*) FROM danmaku_records `+where, args...).Scan(&count)
	return count, err
}

func (s *Store) ClearRecords(ctx context.Context, from, to int64) (int64, error) {
	conditions := make([]string, 0, 2)
	args := make([]any, 0, 2)
	if from != 0 {
		conditions = append(conditions, "ts>=?")
		args = append(args, from)
	}
	if to != 0 {
		conditions = append(conditions, "ts<=?")
		args = append(args, to)
	}
	query := "DELETE FROM danmaku_records"
	if len(conditions) > 0 {
		query += " WHERE " + strings.Join(conditions, " AND ")
	}
	result, err := s.db.ExecContext(ctx, query, args...)
	if err != nil {
		return 0, err
	}
	return result.RowsAffected()
}

func (s *Store) CleanupRetention(ctx context.Context, days, maxRows int) (int64, error) {
	if days < 1 {
		days = 1
	}
	if maxRows < 1 {
		maxRows = 1
	}
	before, err := s.CountRecords(ctx, DanmakuFilter{})
	if err != nil {
		return 0, err
	}
	cutoff := time.Now().Add(-time.Duration(days) * 24 * time.Hour).UnixMilli()
	if _, err := s.db.ExecContext(ctx, `DELETE FROM danmaku_records WHERE ts<?`, cutoff); err != nil {
		return 0, err
	}
	if _, err := s.db.ExecContext(ctx, `DELETE FROM danmaku_records WHERE id IN (SELECT id FROM danmaku_records ORDER BY ts ASC,id ASC LIMIT MAX((SELECT COUNT(*) FROM danmaku_records)-?,0))`, maxRows); err != nil {
		return 0, err
	}
	after, err := s.CountRecords(ctx, DanmakuFilter{})
	return int64(before - after), err
}

func (s *Store) SaveLog(ctx context.Context, entry LogEntry) (int64, error) {
	result, err := s.db.ExecContext(ctx, `INSERT INTO logs(level,category,message,detail,ts) VALUES(?,?,?,?,?)`, entry.Level, entry.Category, entry.Message, nullable(entry.Detail), entry.TS)
	if err != nil {
		return 0, err
	}
	return result.LastInsertId()
}

func (s *Store) ListLogs(ctx context.Context, limit int) ([]LogEntry, error) {
	if limit < 1 || limit > 500 {
		limit = 100
	}
	rows, err := s.db.QueryContext(ctx, `SELECT id,level,category,message,detail,ts FROM logs ORDER BY id DESC LIMIT ?`, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	result := make([]LogEntry, 0)
	for rows.Next() {
		var item LogEntry
		var detail sql.NullString
		if err := rows.Scan(&item.ID, &item.Level, &item.Category, &item.Message, &detail, &item.TS); err != nil {
			return nil, err
		}
		item.Detail = detail.String
		result = append(result, item)
	}
	return result, rows.Err()
}

func (s *Store) GetSetting(ctx context.Context, key string, fallback any, target any) error {
	var raw string
	err := s.db.QueryRowContext(ctx, `SELECT value FROM settings WHERE key=?`, key).Scan(&raw)
	if err == sql.ErrNoRows {
		return json.Unmarshal(mustJSON(fallback), target)
	}
	if err != nil {
		return err
	}
	if err := json.Unmarshal([]byte(raw), target); err != nil {
		return json.Unmarshal(mustJSON(fallback), target)
	}
	return nil
}

func (s *Store) SetSetting(ctx context.Context, key string, value any) error {
	data, err := json.Marshal(value)
	if err != nil {
		return err
	}
	_, err = s.db.ExecContext(ctx, `INSERT INTO settings(key,value) VALUES(?,?) ON CONFLICT(key) DO UPDATE SET value=excluded.value`, key, string(data))
	return err
}

func recordWhere(filter DanmakuFilter) (string, []any) {
	conditions := make([]string, 0, 9)
	args := make([]any, 0, 9)
	if filter.From != 0 {
		conditions = append(conditions, "ts>=?")
		args = append(args, filter.From)
	}
	if filter.To != 0 {
		conditions = append(conditions, "ts<=?")
		args = append(args, filter.To)
	}
	if filter.Source != "" {
		conditions = append(conditions, "source=?")
		args = append(args, filter.Source)
	}
	if filter.Kind != "" {
		conditions = append(conditions, "kind=?")
		args = append(args, filter.Kind)
	}
	if filter.UserName != "" {
		conditions = append(conditions, "user_name LIKE ?")
		args = append(args, "%"+filter.UserName+"%")
	}
	if filter.Keyword != "" {
		conditions = append(conditions, "(text LIKE ? OR gift_name LIKE ?)")
		args = append(args, "%"+filter.Keyword+"%", "%"+filter.Keyword+"%")
	}
	if filter.ActionResult != "" {
		conditions = append(conditions, "action_result=?")
		args = append(args, filter.ActionResult)
	}
	if filter.Matched == "yes" {
		conditions = append(conditions, "matched_rule_id IS NOT NULL")
	} else if filter.Matched == "no" {
		conditions = append(conditions, "matched_rule_id IS NULL")
	}
	if len(conditions) == 0 {
		return "", args
	}
	return "WHERE " + strings.Join(conditions, " AND "), args
}

type rowScanner interface{ Scan(dest ...any) error }

func scanRecord(row rowScanner) (DanmakuRecord, error) {
	var record DanmakuRecord
	var roomID, userID, userName, text, giftName, matchedID, matchedName, failReason, rawJSON sql.NullString
	var giftCount, repeatCount sql.NullInt64
	var giftValue sql.NullFloat64
	err := row.Scan(&record.ID, &record.Source, &roomID, &record.Kind, &userID, &userName, &text, &giftName, &giftCount, &giftValue, &repeatCount, &record.TS, &record.CreatedAt, &matchedID, &matchedName, &record.ActionResult, &failReason, &rawJSON)
	record.RoomID, record.UserID, record.UserName = roomID.String, userID.String, userName.String
	record.Text, record.GiftName, record.MatchedRuleID, record.MatchedRuleName = text.String, giftName.String, matchedID.String, matchedName.String
	record.FailReason, record.RawJSON = failReason.String, rawJSON.String
	if giftCount.Valid {
		record.GiftCount = int(giftCount.Int64)
	}
	if repeatCount.Valid {
		record.RepeatCount = int(repeatCount.Int64)
	}
	if giftValue.Valid {
		record.GiftValue = giftValue.Float64
	}
	return record, err
}

func nullable(value string) any {
	if value == "" {
		return nil
	}
	return value
}

func nullableInt(value int) any {
	if value == 0 {
		return nil
	}
	return value
}

func nullableFloat(value float64) any {
	if value == 0 {
		return nil
	}
	return value
}

func mustJSON(value any) []byte {
	data, _ := json.Marshal(value)
	return data
}
